libqrexec/toml.c

```diff
@@ -0,0 +1,275 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+#include <stddef.h>
+#include <limits.h>
+
+#include "libqrexec-utils.h"
+
+// A trivial parser for a subset of TOML
+static bool qubes_isspace(unsigned char c) {
+    return c == ' ' || c == '\t';
+}
+
+enum toml_type {
+    TOML_TYPE_INVALID, // error
+    TOML_TYPE_BOOL,
+    TOML_TYPE_INTEGER,
+    TOML_TYPE_STRING,
+};
+
+union toml_data {
+    char *string; // allocated with malloc()
+    bool boolean;
+    unsigned long long integer;
+};
+
+static enum toml_type parse_toml_value(
+    const char *file,
+    size_t line,
+    unsigned char *value_to_parse,
+    union toml_data *value) {
+    enum toml_type ty = TOML_TYPE_INVALID;
+    char *end = NULL;
+
+    switch (value_to_parse[0]) {
+    case '0':
+        ty = TOML_TYPE_INTEGER;
+        value->integer = 0;
+        end = (char *)(value_to_parse + 1);
+        break;
+    case '1' ... '9':
+        ty = TOML_TYPE_INTEGER;
+        errno = 0;
+        value->integer = strtoull((char *)value_to_parse, &end, 10);
+        if (errno) {
+            PERROR("%s:%zu: strtoull()", file, line);
+            return TOML_TYPE_INVALID;
+        }
+        break;
+    case 't':
+        ty = TOML_TYPE_BOOL;
+        if (strncmp((char *)value_to_parse, "true", 4) == 0) {
+            value->boolean = true;
+            end = (char *)(value_to_parse + 4);
+            break;
+        } else {
+            LOG(ERROR, "%s:%zu: Unexpected unquoted string", file, line);
+            return TOML_TYPE_INVALID;
+        }
+    case 'f':
+        ty = TOML_TYPE_BOOL;
+        if (strncmp((char *)value_to_parse, "false", 5) == 0) {
+            value->boolean = false;
+            end = (char *)(value_to_parse + 5);
+            break;
+        } else {
+            LOG(ERROR, "%s:%zu: Unexpected unquoted string", file, line);
+            return TOML_TYPE_INVALID;
+        }
+    case '"':
+        LOG(ERROR, "%s:%zu: Double-quoted strings not implemented, use single quotes", file, line);
+        return TOML_TYPE_INVALID;
+    case '\'':
+        ty = TOML_TYPE_STRING;
+        if (value_to_parse[1] == '\'' && value_to_parse[2] == '\'') {
+            LOG(ERROR, "%s:%zu: Triple-quoted strings not implemented", file, line);
+            return TOML_TYPE_INVALID;
+        }
+        end = strchr((char *)value_to_parse + 1, '\'');
+        if (end == NULL) {
+            LOG(ERROR, "%s:%zu: Unterminated quoted string", file, line);
+            return TOML_TYPE_INVALID;
+        } else {
+            size_t to_alloc = (size_t)(end - (char *)value_to_parse);
+            value->string = malloc(to_alloc);
+            if (value->string == NULL) {
+                LOG(ERROR, "%s:%zu: Out of memory copying value", file, line);
+                return TOML_TYPE_INVALID;
+            }
+            memcpy(value->string, value_to_parse, to_alloc - 1);
```

Comment:
```suggestion
            memcpy(value->string, value_to_parse + 1, to_alloc - 1);
```
(copy the string without opening quote, and don't truncate its last char)